### PR TITLE
E2229 Track the time students look at other submissions

### PR DESCRIPTION
**Pull request by Jmanderson1997**
Please set the title of your pull request in the following format: `[Project id/Independent study/Issue fix/...]. project name`, eg., `E904. Integrating courses and assignments on Expertiza with Moodle`

We suggest you including the below information in your pull request:
- A description of the changes proposed in the pull request.
- @mentions of the person or team responsible for reviewing proposed changes.
---
About Expertiza Bot
- If you have any questions about comments given by the expertiza-bot [(example)](https://github.com/expertiza/expertiza/pull/1877#issuecomment-762412918), you could create a comment in your pull request with the format `/dispute [UUID1] [UUID2]` [(example)](https://github.com/expertiza/expertiza/pull/1877#issuecomment-762430057) then the professor and TAs will be notified.
- [Here is a video](https://tinyurl.com/internet-bots) about how to communicate to expertiza-bot.

### CODE REVIEW

File: app/views/popup/reviewer_details_popup.html.erb [hunk 1]
```diff
@@ -6,22 +6,168 @@
...
-      <th align = "right"> Handle:</th>
-      <td> <%= @user.handle %></td>
-    </tr>
-
 
   </table>
 
 </div>
+
+<% def get_link_round_review_time(map_id, round, link) %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:20]: functions should not be defined in views

File: app/views/popup/reviewer_details_popup.html.erb [hunk 2]
```diff
@@ -6,22 +6,168 @@
...
+            <% individual_time = (link_time.end_at.to_i - link_time.start_at.to_i)/60.0 %>
+        <% end %>
+        <% individual_time = individual_time<0? 0: individual_time %>
+        <% sum_time = individual_time + sum_time %>
+    <% end %>
+    <% sum_time %>
+<% end %>
+
+<div id="submissionviewingevent">
+  <% reviewee_list = Array.new %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:36]:
```suggestion
  <% reviewee_list = [] %>
```

use this syntax when you don't initialize the array to anything specific

File: app/views/popup/reviewer_details_popup.html.erb [hunk 3]
```diff
@@ -6,22 +6,168 @@
...
+  <% total_time_list = Array.new %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:37]:
```suggestion
  <% total_time_list = [] %>
```

File: app/views/popup/reviewer_details_popup.html.erb [hunk 4]
```diff
@@ -6,22 +6,168 @@
...
+  <% timeHash = Hash.new %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:38]:
```suggestion
  <% time_hash = {} %>
```

File: app/views/popup/reviewer_details_popup.html.erb [hunk 5]
```diff
@@ -6,22 +6,168 @@
...
+          <% @reviewee_id = map.reviewee_id %>
+          <% link_list = Array.new %>
+          <% @count = 0 %>   <!--- To print revieweeId only once-->
+          <% @dummy = "      " %>    <!---Empty string as a placeholder-->
+          <% @submissionviewingevent_match.each do |submissionviewingevent_entry| %>
+            <% if link_list.include?(submissionviewingevent_entry.link) %><!-- only unique links per round-->
+            <% else %>
+                <% link_list.push(submissionviewingevent_entry.link) %>
+                <tr>
+                <% if @count == 0 %>  <!--- print revieweeId,time, round only for the first row-->
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:72]:
```suggestion
                <% if @count.zero?%>  <!--- print revieweeId,time, round only for the first row-->
```

File: app/views/popup/reviewer_details_popup.html.erb [hunk 6]
```diff
@@ -6,22 +6,168 @@
...
+                 <% @per_round_time << each_link_detail %>
+                <td align="center"><%= sprintf("%.2f",individual_link_time) %></td>
+                <% if @count == 0 %>
+                <td><a onclick="drawChart(<%= @reviewee_id.to_s %>,<%= each_round.to_s %>)">Pie chart view </a></td>
+                <% else %> 
+                <td><%= " "%></td>
+                <%end%>
+                </tr>
+            <% end %>
+            <% @count=@count+1 %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:97]: use each_with_index to manage counters

File: app/views/popup/reviewer_details_popup.html.erb [hunk 7]
```diff
@@ -6,22 +6,168 @@
...
+                <%end%>
+                <td><%= submissionviewingevent_entry.link %></td>
+                <% individual_link_time = get_link_round_review_time(submissionviewingevent_entry.map_id,submissionviewingevent_entry.round,submissionviewingevent_entry.link) %>
+                <% map_total_time = map_total_time + individual_link_time %>
+                 <% each_link_detail = Array.new %>
+                 <% each_link_detail << submissionviewingevent_entry.link %>
+                 <% each_link_detail << individual_link_time %>
+                 <% @per_round_time << each_link_detail %>
+                <td align="center"><%= sprintf("%.2f",individual_link_time) %></td>
+                <% if @count == 0 %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:90]:
```suggestion
                <% if @count.zero? %>
```

File: app/views/popup/reviewer_details_popup.html.erb [hunk 8]
```diff
@@ -6,22 +6,168 @@
...
+    <table width="100%">
+      <tr>
+        <th align="right">Reviewed Team</th>
+        <th align="right">Total Time (mins)</th>  
+        <th align="right">Round </th>
+        <th align="right">Links</th>
+        <th align="right">Time per link (mins)</th>
+        <th></th>
+      </tr>
+    <% @reviewers = Participant.where(user_id: @user.id) %>
```
Comment (johnbumgardner) [inline app/views/popup/reviewer_details_popup.html.erb:52]: why are these using the @ symbol

File: app/helpers/review_mapping_helper.rb [hunk 9]
```diff
@@ -410,6 +410,30 @@ def get_css_style_for_calibration_report(diff)
     css_class
   end
 
+  def get_time_spent_on_review_for_certain_team_for_each_round(map_id)
```
Comment (johnbumgardner) [inline app/helpers/review_mapping_helper.rb:413]: this name is too long. also dont use 'get' in ruby

**Review by johnbumgardner — CHANGES_REQUESTED** [review]

**Comment by LeooHsiang** [comment]
We are building on the Previous team's implementation, E1989. Their solution wrote to the database often and did not optimize storage space to the desired metric.

We have implemented a solution by using the built-in function of pstore to mitigate database interaction and reduce the number of entries in the database significantly.

**Pull request closed without merging**